File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/colorselector/HueView.java
package net.kdt.pojavlaunch.colorselector;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.Nullable;

import net.kdt.pojavlaunch.Tools;

/**
 * 色相を選択するためのビュー。色相グラデーションを表示し、タッチで選択できます。
 */
public class HueView extends View {
    private final Paint blackPaint = new Paint();
    private Bitmap mGamma;
    private HueSelectionListener mHueSelectionListener;
    private float mSelectionHue;
    private float mHeightHueRatio;
    private float mHueHeightRatio;
    private float mWidth;
    private float mHeight;
    private float mWidthThird;

    /**
     * コンストラクタ。ペイントを初期化します。
     */
    public HueView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        blackPaint.setColor(Color.BLACK);
        blackPaint.setStrokeWidth(Tools.dpToPx(3));
    }

    /**
     * 色相選択リスナーを設定します。
     */
    public void setHueSelectionListener(HueSelectionListener listener) {
        mHueSelectionListener = listener;
    }

    /**
     * 色相を設定し、再描画します。
     */
    public void setHue(float hue) {
        mSelectionHue = hue;
        invalidate();
    }

    /**
     * タッチイベントで色相を選択します。
     */
    @SuppressLint("ClickableViewAccessibility")
    @Override
    public boolean onTouchEvent(MotionEvent event) {
        mSelectionHue = event.getY() * mHeightHueRatio;
        invalidate();
        if(mHueSelectionListener != null) mHueSelectionListener.onHueSelected(mSelectionHue);
        return true;
    }

    /**
     * ビューを描画します。色相ビットマップと選択インジケーターを描画します。
     */
    @Override
    protected void onDraw(Canvas canvas) {
        canvas.drawBitmap(mGamma, 0, 0 ,null);
        float linePos = mSelectionHue * mHueHeightRatio;
        canvas.drawLine(0, linePos , mWidthThird, linePos, blackPaint);
        canvas.drawLine( mWidthThird * 2 ,linePos, mWidth, linePos, blackPaint);
    }

    /**
     * サイズ変更時に色相ビットマップを再生成します。
     */
    @Override
    protected void onSizeChanged(int w, int h, int old_w, int old_h) {
        mWidth = w;
        mHeight = h;
        mWidthThird = mWidth / 3;
        regenerateGammaBitmap();
    }

    /**
     * 色相グラデーションのビットマップを生成します。
     */
    protected void regenerateGammaBitmap() {
        if(mGamma != null)
            mGamma.recycle();
        mGamma = Bitmap.createBitmap(getWidth(), getHeight(), Bitmap.Config.ARGB_8888);
        Paint paint = new Paint();
        Canvas canvas = new Canvas(mGamma);
        mHeightHueRatio = 360/ mHeight;
        mHueHeightRatio = mHeight / 360;
        float[] hsvFiller = new float[] {0, 1, 1};
        for(float i = 0; i < mHeight; i++) {
            hsvFiller[0] = i * mHeightHueRatio;
            paint.setColor(Color.HSVToColor(hsvFiller));
            canvas.drawLine(0,i,mWidth, i,paint);
        }
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/view/colorselector/HueView.kt
package com.arata.yukarilauncher.ui.view.colorselector

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import com.arata.yukarilauncher.Tools

/** 色相を選択するための縦方向スライダービュー。全色相のグラデーションを表示する。 */
class HueView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null) : View(context, attrs) {
    /** 選択位置を示す線のペイント。 */
    private val blackPaint = Paint()
    /** 色相グラデーションビットマップ。 */
    private var mGamma: Bitmap? = null
    /** 色相選択リスナー。 */
    private var mHueSelectionListener: HueSelectionListener? = null
    /** 現在選択中の色相値。 */
    private var mSelectionHue = 0f
    /** ピクセル→色相 変換係数。 */
    private var mHeightHueRatio = 0f
    /** 色相→ピクセル 変換係数。 */
    private var mHueHeightRatio = 0f
    /** ビューの幅。 */
    private var mWidth = 0f
    /** ビューの高さ。 */
    private var mHeight = 0f
    /** ビュー幅の3分の1。 */
    private var mWidthThird = 0f

    init {
        blackPaint.apply {
            color = Color.BLACK
            strokeWidth = Tools.dpToPx(3f)
        }
    }

    /** 色相選択リスナーを設定する。 */
    fun setHueSelectionListener(listener: HueSelectionListener?) {
        mHueSelectionListener = listener
    }

    /** 色相値を設定し、ビューを再描画する。 @param hue 設定する色相値（0～360） */
    fun setHue(hue: Float) {
        mSelectionHue = hue
        invalidate()
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        mSelectionHue = event.y * mHeightHueRatio
        invalidate()
        mHueSelectionListener?.onHueSelected(mSelectionHue)
        return true
    }

    override fun onDraw(canvas: Canvas) {
        mGamma?.let { canvas.drawBitmap(it, 0f, 0f, null) }
        val linePos = mSelectionHue * mHueHeightRatio
        canvas.drawLine(0f, linePos, mWidthThird, linePos, blackPaint)
        canvas.drawLine(mWidthThird * 2, linePos, mWidth, linePos, blackPaint)
    }

    override fun onSizeChanged(w: Int, h: Int, oldW: Int, oldH: Int) {
        mWidth = w.toFloat()
        mHeight = h.toFloat()
        mWidthThird = mWidth / 3
        regenerateGammaBitmap()
    }

    /** 色相のグラデーションビットマップを生成する。 */
    protected fun regenerateGammaBitmap() {
        mGamma?.recycle()
        mGamma = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
        val paint = Paint()
        val canvas = Canvas(mGamma!!)
        mHeightHueRatio = 360f / mHeight
        mHueHeightRatio = mHeight / 360f
        val hsvFiller = floatArrayOf(0f, 1f, 1f)
        var i = 0f
        while (i < mHeight) {
            hsvFiller[0] = i * mHeightHueRatio
            paint.color = Color.HSVToColor(hsvFiller)
            canvas.drawLine(0f, i, mWidth, i, paint)
            i++
        }
    }
}